File: p2_procesamiento_por_lotes_multiprogramacion/main.py
import sys
import time
from typing import List

# MY files
from process import Process
import screen
import kb

BATCH: int = 4


def main():
    screen.clear()

    print("Welcome to the Process Simulator\n")
    num_process = get_num_processes()

    batches_res = num_process % BATCH
    batches_cos = num_process / BATCH
    batches = int(batches_cos + (1 if batches_res != 0 else 0))

    h_line(30)
    print(f"Batches: {batches}")
    h_line(30)

    processes = [Process() for _ in range(num_process)]
    execution: List[Process] = []
    finished: List[Process] = []

    screen.clear()

    s_start_time = time.time()

    for i in range(batches):
        p_beg = i * BATCH

        if i == batches - 1 and batches_res > 0:
            p_end = p_beg + batches_res
        else:
            p_end = p_beg + BATCH

        time_elapsed = 0
        time_left = 0
        n = 0

        for j in range(p_beg, p_end):
            execution.append(processes[j])
            time_left += execution[n].get_exe_time()
            n += 1

        j = 0

        # Process printing
        while j < n:
            key = 0
            screen.clear()
            # Batch in execution
            batch_in_exe(i, batches, time_left - time_elapsed)
            finished.append(execution[j])

            # Process of batch in execution
            print("ID\tEstimated Time\tElapsed Time")
            for k in range(j + 1, n):
                process_in_exe(execution[k])

            # Process in Execution
            h_line(70)
            print("Process in execution")
            h_line(70)
            print("ID\tOperation\tEstimated Time\tElapsed Time\tRemaining Time")
            current_process(execution[j])

            p_time = execution[j].get_exe_time()
            key, elapsed = kb.bind(p_time)

            # Time Update
            remaining = p_time - elapsed
            execution[j].add_elapsed_time(elapsed)
            execution[j].set_exe_time(remaining)
            time_elapsed += elapsed

            if key == "w":
                time_elapsed += remaining
                Process.add_p_exe_time(-remaining)
                remaining = 0
                execution[j].set_exe_time(0)
                finished[p_beg + j].set_result("Error")
                j += 1
            elif key == "e":
                execution.insert(n, execution[j])
                execution.pop(j)
                finished.pop(p_beg + j)
            else:
                j += 1

            current_process_times(elapsed, remaining)
            h_line(70)

            # Finiched / Executed Processes
            print("Finished Processes")
            for k, fin in enumerate(finished):
                if k % 4 == 0:
                    h_line(35)
                    finished_batch(k // 4, batches)
                    h_line(35)
                    print("ID\tOperation\tResult")
                finished_processes(fin)
            h_line(35)
            screen.pause()

        execution.clear()

    s_duration = time.time() - s_start_time
    p_duration = time.time() - p_start_time
    h_line(35)
    print(f"Processes Execution Time: {Process.get_p_exe_time()} s")
    print(f"Simulation Execution Time: {round(s_duration, 2)} s")
    print(f"Program Execution Time: {round(p_duration, 2)} s")
    h_line(35)


def get_num_processes() -> int:
    while True:
        try:
            num_process = int(input("Enter the number of processes to execute: "))
            return num_process
        except ValueError:
            print("Invalid input. Please enter a valid integer")
            screen.pause()
            screen.clear()


def h_line(times: int):
    print("-" * times)


def batch_in_exe(i: int, batches: int, time: int):
    h_line(40)
    print(f"Batch in execution: {i+1} of {batches}")
    print(f"Estimated execution time: {time} s")
    h_line(40)


def process_in_exe(p: Process):
    print(f"{p.get_id()}\t{p.get_exe_time()}\t\t{p.get_elapsed_time()}")


def finished_batch(i: int, batches: int):
    print(f"Batch: {i+1} of {batches}")


def finished_processes(p: Process):
    print(f"{p.get_id()}\t{p.get_math_exp()}\t\t{p.get_result()}")


def current_process(p: Process):
    print(f"{p.get_id()}\t{p.get_math_exp()}\t\t{p.get_exe_time()} s", end="")
    sys.stdout.flush()


def current_process_times(elapsed: int, remaining: int):
    print(f"\t\t{elapsed} s\t\t{remaining} s")


if __name__ == "__main__":
    p_start_time = time.time()
    main()

File: p2_procesamiento_por_lotes_multiprogramacion/src/main.rs
mod process;
mod screen;
mod kb;

use std::io::{self, Write};
use std::time::{Duration, Instant};
use std::thread; // For system sleep

const BATCH: usize = 4;

fn main() {
    screen::clear();

    let start = Instant::now();
    let num_process: usize;

    loop {
        println!("Welcome to the Process Simulator\n");
        print!("Enter the number of processes to execute: ");
        io::stdout().flush().expect("Failed to flush stdout");

        let mut n_process = String::new();
        io::stdin()
            .read_line(&mut n_process)
            .expect("Failed to read number");

        num_process = match n_process.trim().parse() {
            Ok(num) => num,
            Err(_) => {
                println!("Invalid input. Please, enter a valid integer");
                screen::pause();
                screen::clear();
                continue; // Continue to the next iteration of the loop
            }
        };

        break;
    }

    let batches_res: usize = num_process % BATCH;
    let batches_cos: usize = num_process / BATCH;
    let batches: usize = batches_cos + if batches_res != 0 { 1 } else { 0 };

    h_line(30);
    println!("Bacthes: {batches}\n");
    h_line(30);


    let mut processes: Vec<process::Process> = Vec::new();
    let mut execution: Vec<process::Process> = Vec::new();
    let mut finished: Vec<process::Process> = Vec::new();

    for _ in 0..num_process {
        processes.push(process::Process::new());
    }

    screen::pause();
    screen::clear();

    let mut p_end;

    for i in 0..batches {
        let p_beg = i * BATCH;

        if i == batches - 1 && batches_res > 0 {
            p_end = p_beg + batches_res;
        } else {
            p_end = p_beg + BATCH;
        }

        let mut n = 0; // Processes counter (lenght) in execution vector
        let mut time_left: u8 = 0; //
        let mut time_elapsed: u8 = 0;

        for j in p_beg..p_end {
            execution.push(processes[j].clone());
            time_left += execution[n].get_exe_time();
            n += 1;
        }

        let mut j = 0;
        let exe_len = execution.len();

        // Process printing
        while j < exe_len {
            screen::clear();
            // Batch in execution
            batch_in_exe(i, batches, time_left - time_elapsed);
            time_elapsed += execution[j].get_exe_time();
            finished.push(execution[j].clone());

            // Process of batch in execution
            println!("ID\tMax Estimated Time");
            for k in j..n {
                process_in_exe(execution[k].get_id(), execution[k].get_exe_time());
            }

            // Process in execution
            h_line(55);
            println!("Process in execution:");
            h_line(55);
            println!("ID\tOperation\tElapsed Time\tBath Time Left");
            current_process(&execution[j]);

            let key = kb::bind();

            match key {
                'w' => {
                    execution[j].set_result("Error");
                    finished[p_beg + j].set_result("Error");
                },
                'e' => {
                    time_elapsed -= execution[j].get_exe_time();

                    execution.splice(exe_len..exe_len , vec![execution[j].clone()]);
                    execution.remove(j);

                    finished.remove(j);
                    continue;
                },
                _ => thread::sleep(Duration::from_secs(*execution[j].get_exe_time() as u64)), // case 'c' too
            }

            j += 1;

            current_process_times(time_elapsed, time_left - time_elapsed);

            // Process executed/finished
            h_line(55);
            println!("Processes finished");
            for (k, fin) in finished.iter().enumerate() {
                if (k) % 4 == 0 {
                    h_line(35);
                    finished_batch((k) / 4, batches);
                    h_line(35);
                    println!("ID\tOperation\tResult");
                }
                finished_processes(fin);
            }

            h_line(35);
            screen::pause(); 
        }

        execution.clear(); // empty the vector
    }

    let duration = start.elapsed();
    h_line(35);
    println!("Processes Execution Time: {} s", unsafe { *process::Process::get_p_exe_time() });
    println!("Program Execution Time: {:?}", duration);
}

fn h_line(times: usize){
    for _ in 0..times {
        print!("-");
    }
    println!();
}

fn batch_in_exe(i: usize, batches: usize, time: u8) {
    h_line(30);
    println!("Batch in execution: {} of {}", i+1, batches);
    println!("Estimated execution time: {} s", time);
    h_line(30);
}

fn process_in_exe(id: &u16, time: &u8) {
    println!("{}\t{}", id, time);
}

fn finished_batch(i: usize, batches: usize) {
    println!("Batch: {} of {}", i+1, batches);
}

fn finished_processes(p: &process::Process) {
    println!("{}\t{}\t\t{}", p.get_id(), p.get_math_exp(), p.get_result());
}

fn current_process(p: &process::Process) {
    print!("{}\t{}", p.get_id(), p.get_math_exp());
    io::stdout().flush().expect("Failed to flush stdout");
}

fn current_process_times(time_1: u8, time_2: u8){
    println!("\t\t{} s\t\t{} s", time_1, time_2);
}
